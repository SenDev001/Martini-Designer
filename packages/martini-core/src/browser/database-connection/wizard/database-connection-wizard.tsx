import { SelectionService } from "@theia/core";
import { OpenerService } from "@theia/core/lib/browser";
import URI from "@theia/core/lib/common/uri";
import { Field, Form, Formik, FormikErrors, FormikProps } from "formik";
import { inject, injectable } from "inversify";
import { noop } from "lodash";
import messages from "martini-messages/lib/messages";
import * as React from "react";
import * as Yup from "yup";
import {
    DatabaseType,
    getTypeLabel, MartiniDatabaseConnectionManager
} from "../../../common/database-connection/martini-database-connection-manager";
import { UpDownLabel } from "../../components/up-down-label";
import { FormEffect, FormRow, OnFormChange, validateSchema } from "../../form/form";
import { Progress } from "../../progress/progress-service";
import { AbstractWizard, AbstractWizardPage, Wizard } from "../../wizard/wizard";
import { WizardContribution } from "../../wizard/wizard-contribution";
import { isDbConnectionSelection } from "../database-connection-contribution";
import { DatabaseConnectionEditor } from "../editor/database-connection-editor";

@injectable()
export class DatabaseConnectionWizardContribution implements WizardContribution {
    @inject(SelectionService)
    private selectionService: SelectionService;
    @inject("Factory<DatabaseConnectionWizard>")
    private factory: () => DatabaseConnectionWizard;

    readonly wizardType = "new";
    readonly id = "database-connection";
    readonly label = messages.database_connection;
    readonly description = messages.create_db_connection;
    readonly iconClass = "martini-icon martini-database-icon";
    readonly keybinding = "ctrlcmd+alt+n d";
    readonly primary = true;

    async createWizard(): Promise<Wizard> {
        return this.factory();
    }

    isVisible() {
        const selection = this.selectionService.selection;
        return isDbConnectionSelection(selection) || (selection instanceof Array && selection[0] === "Databases");
    }

}

@injectable()
export class DatabaseConnectionWizardPage extends AbstractWizardPage {
    existingNames: string[] = [];
    private readonly defaultConfig: DatabaseConnectionConfig = {
        type: DatabaseType.JDBC,
        name: "MyConnection"
    };
    config: DatabaseConnectionConfig = this.defaultConfig;

    render(): React.ReactNode {
        return <CreateDatabaseConnectionForm
            existingNames={this.existingNames}
            defaultConfig={this.defaultConfig}
            onValidate={errors => this.handleValidate(errors)}
            onChange={config => this.handleChange(config)}
        />;
    }

    private handleValidate(errors: FormikErrors<DatabaseConnectionConfig>) {
        this.complete = Object.keys(errors).length === 0;
    }

    private handleChange(config: DatabaseConnectionConfig) {
        this.config = config;
    }
}

export interface DatabaseConnectionConfig {
    type: DatabaseType;
    name: string;
}

export interface CreateDatabaseConnectionFormProps {
    existingNames: string[];
    defaultConfig?: DatabaseConnectionConfig;
    onValidate: (errors: FormikErrors<DatabaseConnectionConfig>) => void;
    onChange: (config: DatabaseConnectionConfig) => void;
}

const databaseConnectionNameRegExp = /^[a-zA-Z0-9\\_-]+$/;

export const CreateDatabaseConnectionForm: React.FC<CreateDatabaseConnectionFormProps> = (
    { existingNames, defaultConfig, onValidate, onChange }) => {

    const schema = React.useMemo(() => Yup.object().shape<Omit<DatabaseConnectionConfig, "type">>({
        name: Yup.string()
            .required()
            .max(30)
            .matches(databaseConnectionNameRegExp, messages.only_alphanumeric)
            .test("unique-name", messages.connection_name_exists, value => !existingNames.includes(value!))
    }), []);
    const validate = (config: DatabaseConnectionConfig) => validateSchema(schema, config, onValidate);

    const handleKeyDown = (e: React.KeyboardEvent, props: FormikProps<DatabaseConnectionConfig>) => {
        if (e.key !== "ArrowUp" && e.key !== "ArrowDown")
            return;
        const types = Object.values(DatabaseType);
        let index = types.indexOf(props.values.type);
        if (e.key === "ArrowUp")
            index = index - 1 < 0 ? types.length - 1 : index - 1;
        else if (e.key === "ArrowDown")
            index = index + 1 >= types.length ? 0 : index + 1;
        const type = types[index];
        props.setValues({
            ...props.values,
            type
        });
        e.preventDefault();
    };

    return <Formik
        initialValues={defaultConfig}
        initialTouched={{
            type: true,
            name: true
        }}
        validate={validate}
        validateOnMount={true}
        onSubmit={noop}
    >
        {props => (
            <Form style={{
                display: "grid",
                gridColumnGap: 10,
                gridTemplateColumns: "max-content 1fr"
            }}>
                <FormEffect deps={[existingNames]} />
                <OnFormChange onChange={onChange} />
                <FormRow name="type" label={messages.type}>
                    <Field as="select" name="type" className="theia-select">
                        {Object.values(DatabaseType).map(type => (
                            <option key={type}
                                value={type}>{getTypeLabel(type)}</option>))}
                    </Field>
                </FormRow>
                <FormRow name="name" label={messages.name} gridTemplateColumns="1fr max-content">
                    <Field
                        type="text"
                        name="name"
                        className="theia-input"
                        autoFocus
                        onKeyDown={(e: React.KeyboardEvent) => handleKeyDown(e, props)}
                    />
                    <UpDownLabel toolTip={messages.pressing_updown_changes_type} />
                </FormRow>
            </Form>)}
    </Formik>;
};

@injectable()
export class DatabaseConnectionWizard extends AbstractWizard {
    readonly title = messages.create_db_connection_title;

    constructor(
        @inject(MartiniDatabaseConnectionManager)
        private readonly dbConnectionMngr: MartiniDatabaseConnectionManager,
        @inject(DatabaseConnectionWizardPage)
        private readonly page: DatabaseConnectionWizardPage,
        @inject(OpenerService)
        private readonly openerService: OpenerService) {
        super();
        this.pages.push(page);
    }

    async init(progress: Progress): Promise<void> {
        progress.report({
            message: messages.fetching_database_connections
        });
        const connections = await this.dbConnectionMngr.getAll();
        this.page.existingNames = connections.map(c => c.name);
        return super.init(progress);
    }

    async performFinish(): Promise<boolean> {
        const uri: URI = new URI(`${DatabaseConnectionEditor.URI_SCHEME}://${encodeURI(this.page.config.name)}`
            + `?type=${this.page.config.type}&new=true`);
        const handler = await this.openerService.getOpener(uri);
        handler.open(uri);
        return true;
    }

}
