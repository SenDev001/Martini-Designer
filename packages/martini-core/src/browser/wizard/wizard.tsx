import { Disposable, DisposableCollection, Emitter, Event } from "@theia/core";
import { injectable } from "inversify";
import messages from "martini-messages/lib/messages";
import * as React from "react";
import { BaseDialog } from "../dialogs/dialogs";
import { Progress } from "../progress/progress-service";

export interface Wizard extends Disposable {
    currentPage: WizardPage | undefined;
    firstPage: WizardPage;
    title: string;
    onCompleteChanged: Event<void>;
    onUpdate: Event<void>;
    hasNext: boolean;
    hasPrevious: boolean;
    initialSize?: { width?: number, height?: number; };

    init(progress: Progress): Promise<void>;

    performFinish(progress: Progress): Promise<boolean>;

    canFinish(): boolean;

    getNextPage(current: WizardPage | undefined): WizardPage | undefined;

    getPreviousPage(current: WizardPage | undefined): WizardPage | undefined;

    render(): React.ReactNode;

    next(): void;

    previous(): void;
}

export interface WizardPage extends Disposable {
    description?: string;
    complete: boolean;
    onCompleteChanged: Event<void>;
    onUpdate: Event<void>;

    render(): React.ReactNode;

    dispose(): void;
}

@injectable()
export abstract class AbstractWizard implements Wizard {
    protected onCompleteChangedEmitter = new Emitter<void>();
    readonly onCompleteChanged = this.onCompleteChangedEmitter.event;
    protected onUpdateEmitter = new Emitter<void>();
    readonly onUpdate = this.onUpdateEmitter.event;

    abstract title: string;
    protected pages: WizardPage[] = [];
    protected history: WizardPage[] = [];
    protected current: WizardPage;
    protected toDispose = new DisposableCollection();

    /**
     * Initializes the wizard. If this method is overridden it should call super when all the pages have been added already.
     *
     * @param progress used to report progress
     */
    async init(progress: Progress): Promise<void> {
        if (this.pages.length) this.current = this.firstPage;
        this.toDispose.pushAll(this.pages);
    }

    abstract performFinish(progress: Progress): Promise<boolean>;

    canFinish(): boolean {
        return (
            this.pages.length === 0 ||
            (this.currentPage !== undefined &&
                this.currentPage.complete &&
                this.getNextPage(this.current) === undefined)
        );
    }

    getNextPage(current: WizardPage | undefined): WizardPage | undefined {
        if (!current) return undefined;
        const index = this.pages.indexOf(current);
        if (index >= 0 && index + 1 < this.pages.length)
            return this.pages[index + 1];
        return undefined;
    }

    getPreviousPage(current: WizardPage | undefined): WizardPage | undefined {
        return this.history.length === 0 ? undefined : this.history[this.history.length - 1];
    }

    render(): React.ReactNode {
        const page = this.currentPage;
        if (page) {
            return (
                <div style={{ display: "grid", overflowY: "hidden" }}>
                    {page.description && (
                        <>
                            <div
                                style={{
                                    marginBottom: 5,
                                    padding: "6px 12px"
                                }}
                            >
                                {page.description}
                            </div>
                            <div
                                style={{
                                    borderTop: "1px solid var(--theia-panel-border)",
                                    width: "100%",
                                    marginBottom: 5
                                }}
                            />
                        </>
                    )}
                    <div style={{ padding: "6px 12px", overflowY: "auto" }}>{page.render()}</div>
                </div>
            );
        }
        return <></>;
    }

    get firstPage() {
        return this.pages[0];
    }

    get currentPage() {
        return this.current;
    }

    get hasNext() {
        return this.pages.length > 1;
    }

    get hasPrevious() {
        return this.hasNext;
    }

    next(): void {
        if (this.current && this.current.complete) {
            const next = this.getNextPage(this.current);
            if (next) {
                this.history.push(this.current);
                this.current = next;
                this.onCompleteChangedEmitter.fire();
                this.onUpdateEmitter.fire();
            }
        }
    }

    previous(): void {
        if (this.current) {
            const previous = this.getPreviousPage(this.current);
            if (previous) {
                this.current = previous;
                this.history.pop();
                this.onCompleteChangedEmitter.fire();
                this.onUpdateEmitter.fire();
            }
        }
    }

    dispose(): void {
        this.toDispose.dispose();
    }
}

@injectable()
export abstract class AbstractWizardPage implements WizardPage {
    protected _complete = false;
    protected onCompleteChangedEmitter = new Emitter<void>();
    readonly onCompleteChanged = this.onCompleteChangedEmitter.event;
    protected onUpdateEmitter = new Emitter<void>();
    readonly onUpdate = this.onUpdateEmitter.event;
    protected toDispose = new DisposableCollection();
    protected _description = "";

    abstract render(): React.ReactNode;

    get complete() {
        return this._complete;
    }

    set complete(complete: boolean) {
        this._complete = complete;
        this.onCompleteChangedEmitter.fire();
    }

    set description(description: string) {
        if (this._description !== description) {
            this._description = description;
            this.onUpdateEmitter.fire();
        }
    }

    get description() {
        return this._description;
    }

    dispose(): void {
        this.toDispose.dispose();
    }
}

export interface WizardDialogProps {
    acceptLabel?: string;
    width?: number;
    height?: number;
}

export class WizardDialog extends BaseDialog<boolean> {
    protected readonly previousBtn: HTMLButtonElement | undefined;
    protected readonly nextBtn: HTMLButtonElement | undefined;
    protected readonly currentPageDisposable = new DisposableCollection();
    protected oldCurrent: WizardPage | undefined;
    private performingFinish = false;
    value = false;

    constructor(protected readonly wizard: Wizard, props?: WizardDialogProps) {
        super({
            title: wizard.title
        });
        const acceptLabel =
            props && props.acceptLabel ? props.acceptLabel : messages.finish;

        if (wizard.hasPrevious) {
            this.previousBtn = this.appendButton(this.createButton(messages.previous));
            this.previousBtn.addEventListener("click", () => this.wizard.previous());
        }
        if (wizard.hasNext) {
            this.nextBtn = this.appendButton(this.createButton(messages.next));
            this.nextBtn.addEventListener("click", () => this.wizard.next());
        }
        this.appendCloseButton(messages.cancel_btn);
        this.appendAcceptButton(acceptLabel);

        if (props && (props.width || props.height)) {
            const style = (this.node.children.item(0) as HTMLDivElement).style;
            style.justifyContent = "flex-start";
            if (props.width) {
                style.maxWidth = `${props.width}px`;
                style.width = `${props.width}px`;
            }

            if (props.height) {
                style.maxHeight = `${props.height}px`;
                style.height = `${props.height}px`;
            }
        }

        this.contentNode.setAttribute(
            "style",
            "justify-content: flex-start; padding-left: 0; padding-right: 0; overflow-y: hidden;"
        );
    }

    protected async init(progress: Progress): Promise<void> {
        if (this.acceptButton)
            this.setEnabled(this.acceptButton, false);
        await this.wizard.init(progress);
        this.wizard.onCompleteChanged(() => this.updateButtons());
        this.wizard.onUpdate(() => this.update());
        this.updateButtons();
        this.update();
    }

    protected doRender(): React.ReactNode {
        this.currentPageDisposable.dispose();
        if (this.wizard.currentPage) {
            this.currentPageDisposable.pushAll([
                this.wizard.currentPage.onCompleteChanged(() => this.updateButtons()),
                this.wizard.currentPage.onUpdate(() => this.update())
            ]);
        }
        return this.wizard.render();
    }

    protected async accept(): Promise<void> {
        if (!this.resolve || this.performingFinish || this.acceptButton?.disabled || !this.wizard.canFinish())
            return;
        try {
            if (this.acceptButton)
                this.setEnabled(this.acceptButton, false);
            this.performingFinish = true;
            this.value = await this.showProgress(undefined, async progress => await this.wizard.performFinish(progress));
        } finally {
            if (this.acceptButton)
                this.setEnabled(this.acceptButton, true);
            this.performingFinish = false;
        }
        if (this.value) {
            this.wizard.dispose();
            return super.accept();
        }
    }

    close(): void {
        this.wizard.dispose();
        super.close();
    }

    protected updateButtons() {
        if (this.isInProgress())
            return;
        if (this.previousBtn) {
            this.setEnabled(
                this.previousBtn,
                this.wizard.getPreviousPage(this.wizard.currentPage) !== undefined
            );
        }

        if (this.nextBtn) {
            this.setEnabled(
                this.nextBtn,
                this.wizard.currentPage !== undefined &&
                this.wizard.currentPage.complete &&
                this.wizard.getNextPage(this.wizard.currentPage) !== undefined
            );
        }
        if (this.acceptButton)
            this.setEnabled(this.acceptButton, this.wizard.canFinish());

        if (this.oldCurrent !== this.wizard.currentPage) {
            this.oldCurrent = this.wizard.currentPage;
            this.update();
        }
    }

    private appendButton(button: HTMLButtonElement): HTMLButtonElement {
        this.controlPanel.appendChild(button);
        button.classList.add("secondary");
        return button;
    }

}
