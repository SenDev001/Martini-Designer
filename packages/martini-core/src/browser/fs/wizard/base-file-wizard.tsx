import { Path } from "@theia/core";
import { Field, Form, Formik, FormikErrors } from "formik";
import { inject, injectable } from "inversify";
import { noop } from "lodash";
import messages from "martini-messages/lib/messages";
import * as React from "react";
import * as Yup from "yup";
import { codeDirRegExp, codeDirResourceRegExp, packageDirRegExp } from "../../../common/fs/file-util";
import { Directory, Resource } from "../../../common/fs/martini-filesystem";
import { UpDownLabel } from "../../components/up-down-label";
import { FormRow, OnFormChange, validateSchemaAsync } from "../../form/form";
import { Progress } from "../../progress/progress-service";
import { AbstractWizard, AbstractWizardPage, Wizard, WizardDialogProps } from "../../wizard/wizard";
import { WizardContribution, WizardType } from "../../wizard/wizard-contribution";
import { FileBrowserDialog, FileBrowserDialogProps } from "../file-browser/file-browser-dialog";
import { ResourceLocationValidator } from "../resource-location-validator";
import { ResourceNameValidator } from "../resource-name-validator";
import { FileWizardHelper } from "./file-wizard-helper";

export interface FileConfig {
    location: string;
    name: string;
}

export interface CreateFileFormProps {
    initialLocation: string;
    initialName: string;
    fileExt: string;
    validateName?: (path: string, name: string) => Promise<true | string>;
    validateLocation?: (parentPath: string) => Promise<true | string>;
    onBrowseLocation?: (location: string) => Promise<string | undefined>;
    onValidate: (config: FileConfig, errors: FormikErrors<FileConfig>) => void;
    onChange: (config: FileConfig) => void;
    nameFieldProps?: NameFieldProps;
}

export interface NameFieldProps {
    onUpDownKey?: (key: "up" | "down") => void;
    upDownLabelTooltip?: string;
}

export const CreateFileForm: React.FC<CreateFileFormProps> = ({
    initialLocation,
    initialName,
    fileExt,
    validateName,
    validateLocation,
    onBrowseLocation,
    onValidate,
    onChange,
    nameFieldProps,
    children
}) => {
    const nameInput = React.useRef<HTMLInputElement>();
    React.useEffect(() => {
        if (nameInput.current) {
            nameInput.current.focus();
            nameInput.current.select();
        }
    }, []);

    const schema = React.useMemo(() => Yup.object().shape<FileConfig>({
        location: Yup.string().required()
            .test("valid-location", "", async function (location) {
                if (!validateLocation || location?.length === 0)
                    return true;
                const result = await validateLocation(location!);
                if (result !== true)
                    return this.createError({ message: result });
                return true;
            }),
        name: Yup.string().required()
            .test("valid-name", "", async function (name) {
                if (!validateName)
                    return true;
                const fullName = fileExt ? name + "." + fileExt : name;
                const result = await validateName(
                    new Path(this.parent.location).toString(),
                    fullName!
                );

                if (result !== true)
                    return this.createError({ message: result });
                return true;
            })
    }), []);

    const validate = async (config: FileConfig) => validateSchemaAsync(schema, config, errors => onValidate(config, errors!));

    return (
        <Formik
            initialValues={{
                location: initialLocation,
                name: initialName
            }}
            initialTouched={{
                location: true,
                name: true
            }}
            validate={validate}
            validateOnMount={true}
            onSubmit={noop}
        >
            {context => {
                const locationInput = React.useRef<HTMLInputElement>();
                const handleBrowse = async () => {
                    if (onBrowseLocation) {
                        const newLocation = await onBrowseLocation(context.values.location);
                        if (newLocation)
                            context.setFieldValue("location", newLocation);
                        locationInput.current?.focus();
                    }
                };

                let locationField: React.ReactNode;

                if (onBrowseLocation) {
                    locationField = <FormRow name="location" label={messages.location} gridTemplateColumns="1fr max-content">
                        <Field name="location" type="text" className="theia-input" innerRef={locationInput} />
                        <input type="button" className="theia-button" value={messages.browse_btn} onClick={handleBrowse} />
                    </FormRow>;
                } else {
                    locationField = <FormRow name="location" label={messages.location}>
                        <Field name="location" type="text" className="theia-input" innerRef={locationInput} />
                    </FormRow>;
                }

                let nameField: React.ReactNode;
                if (nameFieldProps && nameFieldProps.onUpDownKey) {
                    const handleKeyDown = (e: React.KeyboardEvent) => {
                        if (e.key === "ArrowUp") {
                            nameFieldProps.onUpDownKey!("up");
                            e.preventDefault();
                        }
                        else if (e.key === "ArrowDown") {
                            nameFieldProps.onUpDownKey!("down");
                            e.preventDefault();
                        }
                    };
                    nameField = <FormRow name="name" label={messages.name} gridTemplateColumns="1fr max-content">
                        <Field name="name" type="text" className="theia-input"
                            innerRef={nameInput} onKeyDown={(e: React.KeyboardEvent) => handleKeyDown(e)} />
                        <UpDownLabel toolTip={nameFieldProps.upDownLabelTooltip || ""} />
                    </FormRow>;
                } else {
                    nameField = <FormRow name="name" label={messages.name}>
                        <Field name="name" type="text" className="theia-input" innerRef={nameInput} />
                    </FormRow>;
                }
                return <Form
                    style={{
                        display: "grid",
                        gridColumnGap: 10,
                        gridTemplateColumns: "max-content 1fr"
                    }}
                >
                    <OnFormChange onChange={onChange} />
                    {locationField}
                    {nameField}
                    {children}
                </Form>;
            }}
        </Formik>
    );
};

export const FileWizardPageProps = Symbol("FileWizardPageProps");

export interface FileWizardPageProps {
    defaultName?: string;
    fileExt?: string;
}

export type FileWizardPageFactory = (props: Partial<FileWizardPageProps>) => FileWizardPage;

@injectable()
export class FileWizardPage extends AbstractWizardPage {
    config: FileConfig;

    @inject("Factory<FileBrowserDialog>")
    protected readonly browseFileDialogFactory: (props: FileBrowserDialogProps) => FileBrowserDialog;
    @inject(ResourceNameValidator)
    protected readonly nameValidator: ResourceNameValidator;
    @inject(ResourceLocationValidator)
    protected readonly locationValidator: ResourceLocationValidator;
    defaultLocation = "";
    protected nameFieldProps: NameFieldProps;

    constructor(
        @inject(FileWizardPageProps)
        protected readonly props: FileWizardPageProps
    ) {
        super();
    }

    render(): React.ReactNode {
        const initialLocation = this.config ? this.config.location : this.defaultLocation || "";
        const initialName = this.config ? this.config.name : this.props.defaultName || "";
        return (
            <CreateFileForm
                initialLocation={initialLocation}
                initialName={initialName}
                fileExt={this.props.fileExt || ""}
                validateName={(parentPath, name) =>
                    this.nameValidator.validate(parentPath, name)}
                validateLocation={location => this.locationValidator.validate(location)}
                onBrowseLocation={location => this.handleBrowserLocation(location)}
                onValidate={(values, errors) => this.handleValidate(values, errors)}
                onChange={config => this.handleChange(config)}
                nameFieldProps={this.nameFieldProps}
            >
                {this.renderChildren()}
            </CreateFileForm>
        );
    }

    protected renderChildren(): React.ReactNode {
        return undefined;
    }

    protected handleValidate(config: FileConfig, errors: FormikErrors<FileConfig>) {
        this.complete = Object.keys(errors).length === 0;
    }

    protected handleChange(config: FileConfig) {
        this.config = config;
    }

    protected async handleBrowserLocation(location: string) {
        return this.browseFileDialogFactory({
            title: messages.select_directory_title,
            resourceFilter: resource => this.filterResource(resource)
        }).open();
    }

    protected filterResource(resource: Resource): boolean {
        return Directory.is(resource) &&
            !codeDirRegExp.test(resource.location) &&
            !codeDirResourceRegExp.test(resource.location);
    }
}

export const DefaultFileWizardProps = Symbol("DefaultFileWizardProps");

export interface DefaultFileWizardProps extends FileWizardPageProps {
    title: string;
    fileContent?: string;
    dir?: boolean;
}

@injectable()
export class DefaultFileWizard extends AbstractWizard {
    @inject(FileWizardHelper)
    protected readonly wizardHelper: FileWizardHelper;
    title = "";

    constructor(
        @inject(DefaultFileWizardProps)
        protected readonly props: DefaultFileWizardProps,
        @inject(FileWizardPage)
        protected readonly page: FileWizardPage) {
        super();
        this.title = this.props.title;
        this.pages.push(page);
    }

    async init(progress: Progress): Promise<void> {
        this.page.defaultLocation = this.wizardHelper.getTargetDirectoryPath() ||
            await this.wizardHelper.getDefaultTargetDirectory() || "";
        return super.init(progress);
    }

    async performFinish(progress: Progress): Promise<boolean> {
        progress.report({
            message: messages.creating_file
        });
        const config = this.page.config;
        if (!config)
            return false;
        if (this.props.dir !== undefined && this.props.dir)
            return await this.createDir(config);
        return await this.createFile(config, this.getFileContent(config));
    }

    protected async createDir(config: FileConfig): Promise<boolean> {
        await this.wizardHelper.makeDir(config.location, config.name);
        return true;
    }

    protected async createFile(config: FileConfig, content: string): Promise<boolean> {
        await this.wizardHelper.createFile(
            config.location,
            config.name + (this.props.fileExt ? "." + this.props.fileExt : ""),
            content);
        return true;
    }

    protected getFileContent(config: FileConfig): string {
        return this.props.fileContent || "";
    }
}

export class CodeFileWizardPage extends FileWizardPage {
    protected filterResource(resource: Resource): boolean {
        return Directory.is(resource) &&
            (codeDirRegExp.test(resource.location) ||
                codeDirResourceRegExp.test(resource.location));
    }
}

export type DefaultFileWizardFactory = (props: Partial<DefaultFileWizardProps>) => DefaultFileWizard;

@injectable()
export class DefaultCodeFileWizard extends DefaultFileWizard {

    constructor(
        @inject(DefaultFileWizardProps)
        protected readonly props: DefaultFileWizardProps,
        @inject(CodeFileWizardPage)
        protected readonly page: CodeFileWizardPage) {
        super(props, page);
    }

    async init(progress: Progress): Promise<void> {
        await super.init(progress);
        if (packageDirRegExp.test(this.page.defaultLocation))
            this.page.defaultLocation += "/code";
    }
}

export type DefaultCodeFileWizardFactory = (props: Partial<DefaultFileWizardProps>) => DefaultCodeFileWizard;

export interface CodeFileWizardContributionProps {
    title: string;
    defaultName: string;
    fileExt: string;
}

@injectable()
export abstract class AbstractCodeFileWizardContribution implements WizardContribution {
    static readonly DEFAULT_PROPS: WizardDialogProps = {
        acceptLabel: messages.create
    };

    @inject(FileWizardHelper)
    protected readonly wizardHelper: FileWizardHelper;
    @inject("Factory<DefaultCodeFileWizard>")
    protected readonly fileWizardFactory: DefaultCodeFileWizardFactory;

    id: string;
    label: string;
    wizardType: WizardType;

    async createWizard(...args: any[]): Promise<Wizard> {
        return this.fileWizardFactory(this.getFileWizardProps());
    }

    protected getFileWizardProps(): DefaultFileWizardProps {
        return {
            title: messages.create_file_title
        };
    }

    isVisible() {
        const targetDirectoryPath = this.wizardHelper.getTargetDirectoryPath();
        return targetDirectoryPath !== undefined &&
            (codeDirRegExp.test(targetDirectoryPath) ||
                codeDirResourceRegExp.test(targetDirectoryPath) ||
                packageDirRegExp.test(targetDirectoryPath));
    }
}
