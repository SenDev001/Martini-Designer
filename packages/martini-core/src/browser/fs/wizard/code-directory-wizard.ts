import { injectable } from "inversify";
import { AbstractCodeFileWizardContribution, DefaultFileWizardProps } from "./base-file-wizard";
import messages from "martini-messages/lib/messages";

@injectable()
export class CodeDirectoryWizardContribution extends AbstractCodeFileWizardContribution {
    readonly id = "code-directory";
    readonly wizardType = "new";
    readonly label = messages.code_dir_title;
    readonly description = messages.create_new_code_dir;
    readonly iconClass = "martini-icon martini-code-directory-icon";
    readonly keybinding = "ctrlcmd+alt+n c";
    readonly primary = true;
    readonly menuGroup = "5_misc"

    protected getFileWizardProps(): DefaultFileWizardProps {
        return {
            title: messages.create_code_dir_title,
            defaultName: "newDir",
            dir: true
        };
    }
}
