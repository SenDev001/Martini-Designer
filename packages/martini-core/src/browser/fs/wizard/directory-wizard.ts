import { SelectionService } from "@theia/core";
import { inject, injectable } from "inversify";
import messages from "martini-messages/lib/messages";
import { codeDirRegExp, codeDirResourceRegExp } from "../../../common/fs/file-util";
import { isPackageSelection } from "../../package/martini-package-contribution";
import { Wizard } from "../../wizard/wizard";
import { WizardContribution } from "../../wizard/wizard-contribution";
import { DefaultFileWizardFactory } from "./base-file-wizard";
import { FileWizardHelper } from "./file-wizard-helper";

@injectable()
export class DirectoryWizardContribution implements WizardContribution {
    @inject(FileWizardHelper)
    private readonly wizardHelper: FileWizardHelper;
    @inject("Factory<DefaultFileWizard>")
    private readonly fileWizardFactory: DefaultFileWizardFactory;
    @inject(SelectionService)
    private readonly selectionService: SelectionService;

    readonly id = "directory";
    readonly wizardType = "new";
    readonly label = messages.directory;
    readonly description = messages.create_directory;
    readonly iconClass = "martini-icon martini-directory-icon";
    readonly keybinding = "ctrlcmd+alt+n d";
    readonly primary = true;

    async createWizard(): Promise<Wizard> {
        return this.fileWizardFactory({
            title: messages.create_directory_title,
            defaultName: "newDir",
            dir: true
        });
    }

    isVisible() {
        const selection = this.selectionService.selection;

        if (isPackageSelection(selection)) return selection.some(p => p.status !== "UNLOADED" && p.name !== "core");

        const targetDirectoryPath = this.wizardHelper.getTargetDirectoryPath();
        return (
            targetDirectoryPath !== undefined &&
            !codeDirRegExp.test(targetDirectoryPath) &&
            !codeDirResourceRegExp.test(targetDirectoryPath)
        );
    }
}
