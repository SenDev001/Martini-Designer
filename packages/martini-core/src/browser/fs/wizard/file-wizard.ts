import { inject, injectable } from "inversify";
import messages from "martini-messages/lib/messages";
import { codeDirRegExp, codeDirResourceRegExp, packageDirRegExp } from "../../../common/fs/file-util";
import { Wizard } from "../../wizard/wizard";
import { WizardContribution } from "../../wizard/wizard-contribution";
import { DefaultFileWizardFactory } from "./base-file-wizard";
import { FileWizardHelper } from "./file-wizard-helper";

@injectable()
export class FileWizardContribution implements WizardContribution {
    @inject(FileWizardHelper)
    private readonly wizardHelper: FileWizardHelper;
    @inject("Factory<DefaultFileWizard>")
    private readonly fileWizardFactory: DefaultFileWizardFactory;

    readonly id = "file";
    readonly wizardType = "new";
    readonly label = messages.file;
    readonly description = messages.create_file;
    readonly iconClass = "martini-icon martini-file-icon";
    readonly keybinding = "ctrlcmd+alt+n i";
    readonly primary = true;

    async createWizard(): Promise<Wizard> {
        return this.fileWizardFactory({
            title: messages.create_file_title,
            defaultName: "file.txt"
        });
    }

    isVisible() {
        const targetDirectoryPath = this.wizardHelper.getTargetDirectoryPath();
        return (
            targetDirectoryPath !== undefined &&
            !codeDirRegExp.test(targetDirectoryPath) &&
            !codeDirResourceRegExp.test(targetDirectoryPath) &&
            !packageDirRegExp.test(targetDirectoryPath)
        );
    }
}
