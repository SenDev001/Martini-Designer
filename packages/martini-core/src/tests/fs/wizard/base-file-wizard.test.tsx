import * as Enzyme from "enzyme";
import * as Adapter from "enzyme-adapter-react-16";
import { noop } from "lodash";
import * as React from "react";
import * as ReactDOM from "react-dom/test-utils";
import * as Renderer from "react-test-renderer";
import "reflect-metadata";
import { CreateFileForm } from "../../../browser/fs/wizard/base-file-wizard";

Enzyme.configure({ adapter: new Adapter() });

test("CreateFileForm should be rendered", () => {
    const component = Renderer.create(<CreateFileForm
        initialLocation="/examples/code"
        initialName="MyFile"
        fileExt="txt"
        onValidate={noop}
        onChange={noop}
    />);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
});

test("Browsing location should set the input value", async () => {
    const handleBrowse = jest.fn(async () => "/examples/code/test");
    const form = Enzyme.mount(<CreateFileForm
        initialLocation="/examples/code"
        initialName="MyFile"
        fileExt="txt"
        onValidate={noop}
        onChange={noop}
        onBrowseLocation={handleBrowse}
    />);
    await ReactDOM.act(async () => {
        form.find('input[value="Browse..."]').first().simulate("click");
        await new Promise(r => setTimeout(r, 200));
    });
    form.update();
    const input = form.find('input[name="location"]').first();
    expect(handleBrowse).toBeCalledWith("/examples/code");
    expect(handleBrowse).toBeCalledTimes(1);
    expect(input.prop("value")).toBe("/examples/code/test");
});
