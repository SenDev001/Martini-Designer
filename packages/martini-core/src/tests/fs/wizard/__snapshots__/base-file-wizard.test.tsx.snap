// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateFileForm should be rendered 1`] = `
<form
  action="#"
  onReset={[Function]}
  onSubmit={[Function]}
  style={
    Object {
      "display": "grid",
      "gridColumnGap": 10,
      "gridTemplateColumns": "max-content 1fr",
    }
  }
>
  <label
    htmlFor="location"
    onContextMenu={[Function]}
    style={
      Object {
        "marginTop": "3px",
        "verticalAlign": "center",
      }
    }
  >
    Location
  </label>
  <div
    style={
      Object {
        "display": "grid",
        "gridTemplateColumns": "1fr",
        "gridTemplateRows": "max-content max-content",
      }
    }
  >
    <input
      className="theia-input"
      name="location"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="/examples/code"
    />
    <div
      className="form-error-message-wrapper"
    />
  </div>
  <label
    htmlFor="name"
    onContextMenu={[Function]}
    style={
      Object {
        "marginTop": "3px",
        "verticalAlign": "center",
      }
    }
  >
    Name
  </label>
  <div
    style={
      Object {
        "display": "grid",
        "gridTemplateColumns": "1fr",
        "gridTemplateRows": "max-content max-content",
      }
    }
  >
    <input
      className="theia-input"
      name="name"
      onBlur={[Function]}
      onChange={[Function]}
      type="text"
      value="MyFile"
    />
    <div
      className="form-error-message-wrapper"
    />
  </div>
</form>
`;
